pkg/codegen/importer/hcl2.go

```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
 					})
 				}
 			}
-		} else {
+		case *schema.UnionType:
+			discriminatorPropertyName := objectType.Discriminator
+			selectElementType := func() (*schema.ObjectType, bool) {
+				discriminatorValue, ok := obj[resource.PropertyKey(discriminatorPropertyName)]
+				if !ok {
+					// discriminator property is not present, so we can't select a type
+					return nil, false
+				}
+
+				if !discriminatorValue.IsString() {
+					// discriminator property value is not a string,
+					// so we can't select a type from the union mapping
+					return nil, false
+				}
+
+				discriminatorValueText := discriminatorValue.StringValue()
+				correspondingTypeToken, ok := objectType.Mapping[discriminatorValueText]
+				if ok {
+					for _, elementType := range objectType.ElementTypes {
+						// found the type token
+						// match it against the element type which should be an object
+						elementTypeObject, ok := codegen.UnwrapType(elementType).(*schema.ObjectType)
+						if ok {
+							elementTypeParts := strings.Split(elementTypeObject.Token, ":")
+							unionTypeToken := strings.Split(correspondingTypeToken, ":")
+							if len(elementTypeParts) > 1 && len(unionTypeToken) > 1 {
+								typeName := elementTypeParts[len(elementTypeParts)-1]
+								foundTypeName := unionTypeToken[len(unionTypeToken)-1]
+								matchingTypeNames := typeName == foundTypeName
+								moduleName := elementTypeParts[len(elementTypeParts)-2]
+								foundModuleName := unionTypeToken[len(unionTypeToken)-2]
+								matchingModuleNames := moduleName == foundModuleName
+								if matchingTypeNames && matchingModuleNames {
+									return elementTypeObject, true
+								}
+							}
+						}
+					}
+				}
+
+				return nil, false
+			}
+
+			elementType, ok := selectElementType()
+			if ok {
+				for _, p := range elementType.Properties {
+					x, err := generatePropertyValue(p, obj[resource.PropertyKey(p.Name)])
+					if err != nil {
+						return nil, err
+					}
+					if x != nil {
+						items = append(items, model.ObjectConsItem{
+							Key: &model.LiteralValueExpression{
+								Value: cty.StringVal(p.Name),
+							},
+							Value: x,
+						})
+					}
+				}
+			}
+
```

Comment:
I think we need to do something for `!ok`. We can fall back to what we were doing before.